File: Graph/DijkstraAlgorithm/DijkstraAlgorithm.py
import heapq


class Graph:
    def __init__(self):
        self.vertices = set()
        self.edges = defaultdict(list)

    def add_edge(self, start, end, weight):
        self.vertices.add(start)
        self.vertices.add(end)
        self.edges[start].append((end, weight))

    def dijkstra(self, start):
        distances = {vertex: float('infinity') for vertex in self.vertices}
        distances[start] = 0
        priority_queue = [(0, start)]

        while priority_queue:
            current_distance, current_vertex = heapq.heappop(priority_queue)

            if current_distance > distances[current_vertex]:
                continue

            for neighbor, weight in self.edges[current_vertex]:
                distance = current_distance + weight

                if distance < distances[neighbor]:
                    distances[neighbor] = distance
                    heapq.heappush(priority_queue, (distance, neighbor))

        return distances


# Пример использования
graph = Graph()
graph.add_edge("A", "B", 1)
graph.add_edge("A", "C", 4)
graph.add_edge("B", "C", 2)
graph.add_edge("B", "D", 5)
graph.add_edge("C", "D", 1)
graph.add_edge("D", "A", 7)

start_vertex = "A"
result = graph.dijkstra(start_vertex)
print(f"Shortest distances from vertex {start_vertex}: {result}")

File: Graph/DijkstraAlgorithm/DijkstraAlgorithm.cpp
#include <iostream>
#include <unordered_map>
#include <unordered_set>
#include <queue>
#include <vector>
#include <climits>

class Graph {
private:
    std::unordered_set<char> vertices;
    std::unordered_map<char, std::vector<std::pair<char, int>>> edges;

public:
    void addEdge(char start, char end, int weight) {
        vertices.insert(start);
        vertices.insert(end);
        edges[start].emplace_back(end, weight);
    }

    std::unordered_map<char, int> dijkstra(char start) {
        std::unordered_map<char, int> distances;
        for (char vertex : vertices) {
            distances[vertex] = INT_MAX;
        }
        distances[start] = 0;

        std::priority_queue<std::pair<int, char>, std::vector<std::pair<int, char>>, std::greater<std::pair<int, char>>> pq;
        pq.push({0, start});

        while (!pq.empty()) {
            int currentDistance = pq.top().first;
            char currentVertex = pq.top().second;
            pq.pop();

            if (currentDistance > distances[currentVertex]) {
                continue;
            }

            for (const auto& neighbor : edges[currentVertex]) {
                int distance = currentDistance + neighbor.second;
                if (distance < distances[neighbor.first]) {
                    distances[neighbor.first] = distance;
                    pq.push({distance, neighbor.first});
                }
            }
        }

        return distances;
    }
};

int main() {
    Graph graph;
    graph.addEdge('A', 'B', 1);
    graph.addEdge('A', 'C', 4);
    graph.addEdge('B', 'C', 2);
    graph.addEdge('B', 'D', 5);
    graph.addEdge('C', 'D', 1);
    graph.addEdge('D', 'A', 7);

    char startVertex = 'A';
    std::unordered_map<char, int> result = graph.dijkstra(startVertex);

    std::cout << "Shortest distances from vertex " << startVertex << ":\n";
    for (const auto& entry : result) {
        std::cout << entry.first << ": " << entry.second << "\n";
    }

    return 0;
}